### ISSUE TITLE
[Issue]: Pytorch runtime error

### ISSUE THREAD

**Issue by shifrin8101** [ROCm/ROCm#3840]
### Problem Description

Hello, my GPU output is:
03:00.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Navi 33 [Radeon RX 7700S/7600/7600S/7600M XT/PRO W7600] (rev c0)

I installed with pytorch installation guide.
I have the following error:
ocBLAS error: Cannot read /home/mark/PycharmProjects/WSN-MA/venv/lib/python3.12/site-packages/torch/lib/rocblas/library/TensileLibrary.dat: Illegal seek

do not know how to proceed?

### Operating System

OS: NAME="Ubuntu" VERSION="24.04.1 LTS (Noble Numbat)"

### CPU

CPU:  model name	: Intel(R) Core(TM) i5-14400

### GPU

AMD Radeon Pro W6800

### ROCm Version

ROCm 5.7.1

**Comment by harkgill-amd** [comment]
Hi @shifrin8101, while your GPU is not supported, you can try to set the environment variable `export HSA_OVERRIDE_GFX_VERSION=10.3.0` to mimic a supported architecture. This may allow you to workaround the rocBLAS error but please note that without a supported GPU, it will be difficult to debug any further issues that may arise.

**Comment by harkgill-amd** [comment]
@shifrin8101, did you get a chance to try out the workaround?

**Comment by harkgill-amd** [comment]
Closing this issue out. @shifrin8101, if you are still encountering issues after setting the HSA Override environment variables, please leave a comment and I will re-open this ticket. Thanks!

**Comment by shifrin8101** [comment]
Hi, I have not had a chance to get to my AMD desktop yet. Will have it in
the next few days, hopefully.
Will let you know, so it would take some patience please :)
Thanks!

‫בתאריך יום ו׳, 18 באוק׳ 2024 ב-17:02 מאת ‪harkgill-amd‬‏ <‪
***@***.***‬‏>:‬

**Comment by harkgill-amd** [comment]
Upon re-examination, your setup has a Navi33 which is `gfx1102`. The HSA Override method works effectively on older GPUs such as Navi22 as the ISAs were identical (gfx1036 ISA is identical to gfx1030). With Navi33/gfx1102, the most relevant override would be to Navi31/gfx1100, but this still isn't guaranteed to work and not recommended as there are known differences between these two chips. 

I'll confirm if there are any other workarounds present for Navi33. In the meantime, could you please provide the complete sample program you are running so we can test this on our end as well?

**Comment by shifrin8101** [comment]
Here, this the simplest example which results  the latest error:
import torch

def get_device():
    if torch.cuda.is_available():
        #device =  torch.cuda.get_device_name(0)
        device = torch.device(0)
        print(f'device name [0]:', torch.cuda.get_device_name(0), ' GPU is available')
    else:
        device = torch.device("cpu")
        print("GPU is not available. Using CPU.")
    return device


device = get_device()
print(device)

torch.autograd.set_detect_anomaly(True)

w1_init = torch.randn(2, 2)
try:
    W1=w1_init.clone().detach().to(device).requires_grad_(True)
except Exception as e:
    print(f"An error occurred: {e}")

finished_sample=1

**Comment by shifrin8101** [comment]
for some reason the indentation wasn't kept, sorry

**Comment by harkgill-amd** [comment]
I wasn't able to reproduce the error on my end with a Navi33, see the output of the sample below.
```
device name [0]: AMD Radeon RX 7600  GPU is available
cuda:0
```
Could you please try to run the sample on the rocm/pytorch docker image? You can launch a container with the image by utilizing the following commands:
```
docker pull rocm/pytorch:latest
docker run -it --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
--device=/dev/kfd --device=/dev/dri --group-add video \
--ipc=host --shm-size 8G rocm/pytorch:latest
```
Also, try setting the `HIP_VISIBLE_DEVICES` environment variable at the beginning of your script with `os.environ['HIP_VISIBLE_DEVICES']='0'`. This will ensure that your dGPU is being targeted during execution.

**Comment by shifrin8101** [comment]
Hi,
Will try this during the following week, thanks a lot, meanwhile

**Comment by shifrin8101** [comment]
Hi sorry for the delay, I never tried anything with docker image before.

It actually writes:
sudo docker pulldocker pull rocm/pytorch:latest
docker run -it --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
--device=/dev/kfd --device=/dev/dri --group-add video \
--ipc=host --shm-size 8G rocm/pytorch:latest
latest: Pulling from rocm/pytorch
Digest: sha256:54422bee895f9e44bc5257ab03011aae532c5b7cfa39dda00a3000c46db81239
Status: Image is up to date for rocm/pytorch:latest
docker.io/rocm/pytorch:latest
docker: unknown server OS: .
See 'docker run --help'.

So I am not sure I did the right thing. Also pleas note I am not running the pycharm with sudo (it cause me problems)
So getting back to you...

Thanks

**Comment by shifrin8101** [comment]
sorry for being a nube but I failed again with the docker:
mark$:08:04:36:~/PycharmProjects/WSN-MA > sudo dockersudo docker run -it --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
--device=/dev/kfd --device=/dev/dri --group-add video \
--ipc=host --shm-size 8G rocm/pytorch:latest
[sudo] password for mark: 
root@279f841b1155:/var/lib/jenkins# docker run sample_prog.py
bash: docker: command not found

while I generally understand that this all means I have some VM deployed which mimics a system with pytorch I had zero experience and the output I see already deviates from what is written in help pages on docker...

**Comment by shifrin8101** [comment]
OK I just created new file - same code.
p1.py looks like this:
import torch
import os
os.environ['HIP_VISIBLE_DEVICES']='0'
def get_device():
    if torch.cuda.is_available():
        #device =  torch.cuda.get_device_name(0)
        device = torch.device(0)
        print(f'device name [0]:', torch.cuda.get_device_name(0), ' GPU is available')
    else:
        device = torch.device("cpu")
        print("GPU is not available. Using CPU.")
    return device


device = get_device()
print(device)

torch.autograd.set_detect_anomaly(True)

w1_init = torch.randn(2, 2)
try:
    W1=w1_init.clone().detach().to(device).requires_grad_(True)
    print(W1)
except Exception as e:
    print(f"An error occurred: {e}")

finished_sample=1

running at the docker:

python p1.py
root@416198dec828:/var/lib/jenkins# python p1.py
device name [0]: AMD Radeon™ RX 7600 XT  GPU is available
cuda:0
An error occurred: HIP error: invalid device function
HIP kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing AMD_SERIALIZE_KERNEL=3.
Compile with `TORCH_USE_HIP_DSA` to enable device-side assertions.

running in my machine just made it 
device name [0]: AMD Radeon RX 7600 XT  GPU is available
cuda:0

Process finished with exit code 139 (interrupted by signal 11:SIGSEGV)


so apparently something is fundamentally wrong. (Consider getting a better GPU? which one?)

hope this provides some useful info about the situation.
thanks!

**Comment by harkgill-amd** [comment]
Could you please provide the dmesg output after you hit the error.

**Comment by shifrin8101** [comment]
[dmesg_output.txt](https://github.com/user-attachments/files/17836868/dmesg_output.txt)

OK, that one produced really a lot, here it is, attached, really hope it would be useful!

**Comment by YumingChang02** [comment]
@harkgill-amd may i ask if the ISA disagree is across 780m  7600(xt) 7700/7800xt? is there some insight to which rdna3 gpus are ISA compatable with NAVI31/gfx1100?

**Comment by harkgill-amd** [comment]
@shifrin8101, I was finally able to reproduce the `Invalid device function` error with a 7600XT using a different workload. Did some digging and it is due to the gfx1102 architecture not being supported, see the relevant logs below.

```
1:hip_fatbin.cpp           :117 : 6041234758 us: [pid:840   tid:0x78e4cb2a4740] Missing CO for these ISAs -
:1:hip_fatbin.cpp           :120 : 6041234760 us: [pid:840   tid:0x78e4cb2a4740]      amdgcn-amd-amdhsa--gfx1102
:3:hip_platform.cpp         :715 : 6041234763 us: [pid:840   tid:0x78e4cb2a4740] init: Returned hipErrorNoBinaryForGpu : continue parsing remaining modules
:1:hip_fatbin.cpp           :276 : 6041235417 us: [pid:840   tid:0x78e4cb2a4740] Cannot find CO in the bundle /opt/rocm-6.2.3/magma/lib/libmagma.so for ISA: a
```
Overriding to gfx1100 with `export HSA_OVERRIDE_GFX_VERSION=11.0.0` worked around this issue and would be your best bet despite the ISAs not being identical. There may be certain low level instructions that are missing due to the differences between the versions which can cause crashes. However, in most cases workloads should run without any issues. You can also set this variable locally with the following so it only affects the current Python process
```
import os
os.environ['HSA_OVERRIDE_GFX_VERSION'] = '11.0.0'
```

@YumingChang02 Here's a list that answers your question https://salsa.debian.org/rocm-team/community/team-project/-/wikis/Supported-GPU-list#architecture-notes. The ISAs for gfx1100, gfx1101 and gfx1102 all have differences but in cases such as this, overriding to gfx1100 is the only course that can provide usability.

**Comment by shifrin8101** [comment]
Thanks a lot for this great assistance! 
So far it is successfully running.
I will check the situation again after some runtime will pass. Will let know here if new problems emerge, otherwise will let know that the solution to the problem was found.

**Comment by harkgill-amd** [comment]
Hi @shifrin8101, I'll close this issue out as the workload runs correctly with the `HSA_OVERRIDE_GFX_VERSION` environment variable set. If you do have any other questions or updates, feel free to leave a comment on this thread and I'll re-open this issue.
